### ISSUE TITLE
Push doesn't seem to work

### ISSUE THREAD

**Issue by mark-hahn** [jacogr/atom-git-control#7]
I am probably misunderstanding something but push seems broken.  I do a commit, the list is empty, then I try a push and nothing happens.  The little cloud icon shows an up arrow with the number 2 so it knows a push is needed.

## Versions

* **Atom:**       0.165.0
* **Atom-Shell:** 0.19.4
* **OS:**         Microsoft Windows 8.1 Pro with Media Center
* **Misc**
    * apm  0.116.0
    * npm  1.4.4
    * node 0.10.33
    * python 2.7.8
    * git 1.7.11.msysgit.1
    * visual studio 2013

---

<small>This report was created in and posted from the Atom editor using the package `bug-report` v0.5.3.</small>

**Comment by KapuzenSohn** [comment]
I just Testet it and when I commit something the cloud icon don´t shows up. When I now press on Push it pushes the commit like expected.

```
Atom: 0.166.0-25b44ce (builded on the master branch)
OS: Linux Mint 17
apm: 0.116.0
npm: 1.4.4
node: 0.10.33
python: 2.7.6
git 1.9.1
```

**Comment by jacogr** [comment]
@mark-hahn Definitely sounds like an issue. The "cloud icon" and "push button" should be in-sync. I'll need to walk through this.

@KapuzenSohn Bit harder to trace yours, but I'll walk through the scenarios to see if I can come up with anything related.

**Comment by mark-hahn** [comment]
>  The "cloud icon" and "push button" should be in-sync.

The push button is enabled so the display is in sync.  The push button just
doesn't do anything.

Let me know how I can help.

**Comment by jacogr** [comment]
Ouch. Weird. It should push. Don't have any bright ideas immediately, but I'll look through and see what I can come up with.

**Comment by mark-hahn** [comment]
Am I supposed to select something like a branch before clicking on push?  I
tried selecting everything with no luck.

**Comment by jacogr** [comment]
It really shouldn't be any different from what you would normally do -

0. Make changes
1. Select files to commit
2. Commit
3. Push

Between 2 & 3 the push button becomes active, clicking it should just push. (No real feedback apart from the little log window, but push should push.)

**Comment by mark-hahn** [comment]
OK, that's what I'm trying to do.  The log last line says ...

    > git -c push.default=simple push --porcelain

**Comment by jacogr** [comment]
So it does try to do the push. (There should probably be some indicator that it is busy)

Once the push is completed, it should either (a) report the error (if anything fails) or (b) do a status and render the push button inactive.

**Comment by mark-hahn** [comment]
I just committed, clicked on push, and waited several minutes.  That log
line just sits there.  No error and the push doesn't happen and the push
button stays enabled.  Nothing appears in my github repo.

**Comment by jacogr** [comment]
![image](https://cloud.githubusercontent.com/assets/1424473/5610285/3033e4de-94bd-11e4-9f62-f247a908d1fa.png)

It is actually quick, nothing more than a second or two. So for some reason it doesn't return from the push. I'll need to see how I can get more info from that part of the process, basically calling into git-promise

**Comment by mark-hahn** [comment]
I was playing around and I did a compare and it failed with this error ...

    > git --no-pager diff
Cannot read property 'push' of undefined

There is something in my setup or repo making this happen. You can get my
repo at https://github.com/mark-hahn/node-ide

**Comment by jacogr** [comment]
Thank you, will grab it and see if I can reproduce.

If you do it via the command-line, does a "git push" work?

**Comment by mark-hahn** [comment]
I found the problem.  When git-control wasn't working I used SourceTree and
it worked fine.  I just now tried `git push` on the command line and it
asked for a login.  So it must be doing the same thing when I click the
push button.  It just doesn't show me any indication that is happening.

SourceTree must be keeping it's own login to use.  This problem is
surprising since I have used the command line recently with no problem.

I'll fix this now.  I'll google how to set the login in git permanently.
(grin)

Sorry for your trouble.  Consider this a feature request to show everything
the git command puts out in the log.  Specifically the login request.

**Comment by jacogr** [comment]
Ahh, ok.

I think the prompt info is not returned from the git package I use, I'll figure out a way around this. Going to try and get it reproduced so I can work on a way of addressing this.

**Comment by mark-hahn** [comment]
Yeah.  I think others may run into it also.  Maybe there is some test that
can be done when git-control first loads to check whether origin can be
reached.

Did you ever see
https://discuss.atom.io/t/shout-out-for-package-git-control/13978 ?

**Comment by jacogr** [comment]
Saw the thread now, thank you :)

Was actually inspired quite a bit via SourceTree, certainly replaced most of my use of the tool as well. (There are some stuff which I still do in SourceTree, eventually I hope to get all my use in here.)

**Comment by mark-hahn** [comment]
> Saw the thread now, thank you :)

I love sourcetree which must be why I like yours.

I'm not sure why I prefer using your in-atom solution instead of just
clicking outside to sourcetree.  I did a package called web-browser which
is a very rudimentary browser.  When I was chatting about doing it everyone
said "why would you use that instead of just clicking on a real browser?".
I didn't have an answer.  But that package got over a thousand downloads in
a month or two.

>  I hope to get all my use in here

If you could tag a commit then you would have everything needed to publish
a package.

I have a bunch of feature requests for you which I will enter when I get
the time.  Simple stuff like ctrl-enter doing the commit when putting in
the commit message.  When I'm done with this massive package I'm on right
now I'll submit some PRs.

**Comment by jacogr** [comment]
Ok, no closer to actually solving the "prompt issue". Basically using the git-promise package (very small) as a wrapper around git via shellex. This one doesn't currently handle inputs well/at all.

Looking at my options here, will either just wrap git myself or PR git-promise, however just need to test and find something that works first.

**Comment by mark-hahn** [comment]
I wouldn't spend much effort on this. Something in the docs might be nice.

**Comment by cristianszwarc** [comment]
Same issue here. My origin is a private bitbucket and I'm using a private key, git may be executed from another system user and therefore not using the correct ssh key? 

steps: 
* command line: git status: nothing to commit (working directory clean)
* small change on a file
* git control detects the change, able to commit
* click on push after the commit and got same msgs than mark-hahn
* git status on command line shows all is ok (nothing new to commit again, 1 commit ahead)
* git push origin master from command line works fine

**Comment by jacogr** [comment]
@cristianszwarc Thank you. I don't think it is due to being executed as somebody else, rather I don't believe the environment gets propagated properly.

I'll get the same setup as you going and try to see what is up.

**Comment by hellerbarde** [comment]
[How to properly set up credential stores on Windows and Linux](https://github.com/jacogr/atom-git-control/issues/42#issuecomment-121604138)
Hope that helps anyone stumbling across this. 

Is there anything we can do about notifying the user about this problem or asking them for a password?

**Comment by pradyunsg** [comment]
Bump.

How about mentioning/linking https://github.com/jacogr/atom-git-control/issues/42#issuecomment-121604138 in the README for the time-being. It'll be easier to discover there...
